File: cpp/questions/searchMatrix.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rs = matrix.size();
        int cs = matrix[0].size();
        int s = 0;
        int e = rs * cs - 1;
        while (s < e) {
            int m = (s + e) / 2;
            int v = matrix[m/cs][m%cs];
            if (v == target) {
                return true;
            } else if (v > target) {
                e = m - 1;
            } else {
                s = m + 1;
            }
        }
        return matrix[s/cs][s%cs]==target;
    }
};

File: python/questions/searchMatrix.py
from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        lr = len(matrix)
        lc = len(matrix[0])
        s = 0
        e = lr
        while s < e:
            m = (s + e) // 2
            cv = matrix[m][lc-1]
            if cv == target:
                return True
            elif cv > target:
                e = m
            else:
                s = m + 1
        if s >= lr or target < matrix[s][0]:
            return False
        r = s
        s = 0
        e = lc
        while s < e:
            m = (s + e) // 2
            cv = matrix[r][m]
            if cv == target:
                return True
            elif cv > target:
                e = m
            else:
                s = m + 1
        return matrix[r][s] == target
